### ISSUE TITLE
DeepWalk embedding and computational performance needs improvement

### ISSUE THREAD

**Issue by MrChristophRivera** [dmlc/dgl#2807]
## 🐛 Bug

We compared DGLS's DeepWalks implementation to the implementation of [KarateClub](https://github.com/benedekrozemberczki/karateclub) on Zachary's Karate club and found that it performed 3 orders of magnitude slower and had embeddings that were ~1/2 as effective on a ranking task. 

## To Reproduce

Please see the attached notebook html for details. 
Briefly, we: 
1. Encoded Zachary's Karate Club as a text file. 
1. Ran DGL DeepWalk from jupyter notebook using python subprocess. Commands such as `python3 deepwalk.py --data_file karate.txt --output_emb_file karate_emb.txt --only_cpu --window_size 5 --num_walks 10 --negative 5 --num_threads 1 --dim 2` were run.  
1. We scanned a range of hyperparameters. Walk Length: [5, 10, 20, 40, 80], Learning Rate:[0.2, 0.1, 0.05, 0.01]
2. The following hyperparameters were kept as follows: Negatives 4, embedding dimension 2, num_walks 10
3. For each setting we computed the embedding 10 times. 
4. Using the embedding we computed the cosine similarity between a source node (node 0 typically) and other nodes in the graph. 
5. We used the cosine similarity, to rank the nodes and compute Hits@10 for inclusion in the first community. 
6. We repeated the same using KarateClub DeepWalk.

It took DGL DeepWalk took on average 2378.95 seconds per 10 runs, whereas KarateClub takes 1.34 seconds per 10 runs.  

## Expected behavior
We expect the embedding performance to be similar to KarateClub and the computation time to be far less. 

## Environment

 - DGL Version (e.g., 1.0): 0.6.0post1
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): Pytorch 1.8.0
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): conda install
 - Build command you used (if compiling from source):
 - Python version: 3.7.10
 - CUDA/cuDNN version (if applicable): NA
 - GPU models and configuration (e.g. V100): NA
 - Any other relevant information:
We are using the newest version of the pytorch docker. 

## Additional context

Here is a graph showing the relative performance 
![image](https://user-images.githubusercontent.com/4410683/113348517-8918b600-92eb-11eb-8c9d-2308d70e8c9f.png)

**Comment by jermainewang** [comment]
Thanks for the report. We are investigating the problem and will back to you soon.

**Comment by BarclayII** [comment]
Could you attach the notebook so we can see how you did it?  Thanks!

**Comment by VoVAllen** [comment]
Hi,

The default settings in deepwalk examples is designed for large dataset, therefore it will generate a large negative sample table before the actual training process. For small dataset such as karate, you can use `fast_neg` argument to generate negative samples on the fly. For example `python deepwalk.py --data_file karate.txt --fast_neg --only_cpu`

**Comment by VoVAllen** [comment]
On dgl with `fast_neg` setting, it took abot 0.03s per run on my machine on Karate Club. Feel free to reopen this if you have any questions.
